examples/async/dspy_async_impl.py

```diff
@@ -0,0 +1,62 @@
+import dspy
+from concurrent.futures import ThreadPoolExecutor
+import os
+import asyncio
+import time
+
+# Ideally:
+os.environ["DSP_CACHEBOOL"] = "false"
+lm = dspy.LM("gpt-4o-mini", cache=False)
+dspy.settings.configure(lm=lm, async_mode=True)
+
+
+async def gather_with_concurrency(n, *coros):
+    semaphore = asyncio.Semaphore(n)
+
+    async def sem_coro(coro):
+        async with semaphore:
+            return await coro
+
+    return await asyncio.gather(*(sem_coro(c) for c in coros))
+
+
+async def async_main(num_items=200):
+    module = dspy.ChainOfThought("question -> answer")
+
+    start_time = time.time()
+    tasks = [module(question=f"What is 1 + {i}?") for i in range(num_items)]
+    await gather_with_concurrency(20, *tasks)
+    end_time = time.time()
+
+    print(f"Total time (async): {end_time - start_time} seconds")
+
+
+def thread_main(num_items=200):
+    module = dspy.ChainOfThought("question -> answer")
+
+    start_time = time.time()
+    with ThreadPoolExecutor(max_workers=20) as executor:
+        list(
+            executor.map(
+                lambda x: module(question=x),
+                [f"What is 1 + {i}?" for i in range(num_items)],
+            )
+        )
+    end_time = time.time()
+
+    print(f"Total time (threaded): {end_time - start_time} seconds")
+
+
+if __name__ == "__main__":
+    num_items = 500
+    dspy.settings.configure(async_mode=False)
+    thread_main(num_items=num_items)
+    dspy.settings.configure(async_mode=True)
+    # print("Sleeping for 60 seconds")
+    # time.sleep(60)
+    asyncio.run(async_main(num_items))
+
+# edge cases I can think of:
+# - multiple modules in the chain
+# - async gets stuck/fails
+# - what happens to settings inside each async module?
```

Comment:
These are great questions.

Threading in DSPy needs care, because of the singleton dspy.settings. We handle that carefully when launching threads, e.g. see the PR #1690 for how this worked before and how it will work after the PR is merged.

How is async thinking about this?

Reply:
These were Isaac's 😅

Reply:
Is the async execution of mutiple modules not just up to the implementation of the parent module?